contributors/design-proposals/storage/csi-snapshot.md

```diff
@@ -0,0 +1,316 @@
+	// kubelet's host machine and then exposed to the pod.
+	// More info: https://kubernetes.io/docs/concepts/storage/volumes#awselasticblockstore
+	// +optional
+	AWSElasticBlockStore *AWSElasticBlockStoreVolumeSnapshotSource `json:"awsElasticBlockStore,omitempty"`
+        // ... etc. for other snapshot types
+}
+
+```
+
+Here is an example of the VolumeSnapshotDataSource for Amazon EBS snapshots.
+
+```
+
+type AWSElasticBlockStoreVolumeSnapshotSource struct {
+	// Unique id of the persistent disk snapshot resource. 
+	SnapshotID string `json:"snapshotId"`
+}
+
+```
+
+#### The `DataSource` Object in PVC
+
+Add a new field into PVC to represent the source of the data which is prepopulated to the provisioned volume. Possible source types may include the following:
+
+    * Snapshot: restore snapshot to a new volume
+    * PersistentVolumeClaim: clone volume which is represented by PVC
+    * GitRepository: copy data from a git repo
+
+```
+
+type PersistentVolumeClaimSpec struct {
+        // If specified, volume will be prepouplated with data from the DataSource
+        // +optional
+        PVCDataSource DataSource
+}
+
+type DataSource struct {
+        SnapSource SnapshotSource
+}
+
+type SnapshotSource struct {
+        // Name of a VolumeSnapshot
+        Name string
+}
+
+```
+Note: SnapshotSource will be added in this phase. PVCSource will be added next.
+
+
+#### The `SnapshotClass` Object
+
+A new SnapshotClass API object will be added to avoid mixing parameters between snapshots and volumes. Each CSI Volume Driver can have its own default SnapshotClass. If SnapshotClass is not provided, a default will be used. It allows to add new parameters for snapshots.
+
+```
+
+type SnapshotClass struct {
+	// This value may not be empty.
+	Provisioner string
+
+	// Parameters holds parameters for the provisioner.
+	Parameters map[string]string
+}
+
+```
+
+#### Add `DataSource` to `StorageClass`
+
+DataSource will be added as an optional parameter in `StorageClass` to allow volumes belonging to the same `StorageClass` to be created with prepopulated data.
+
+```
+
+type StorageClass struct {
+        // If specified, volume will be prepouplated with data from the DataSource
+        // +optional
+        PVCDataSource DataSource
+}
+
+```
+
+
+### CSI External Snapshot Controller
+* External snapshotter is responsible for creating/deleting snapshots and binding snapshot and snapshotData objects.
+
+* External snapshotter is running in the sidecar along with external-attacher and external-provisioner for each CSI Volume Driver.
+
+* For dynamically created snapshot, it should have a SnapshotClass associated with it. User can explicitly specify a SnapshotClass in the VolumeSnapshot API object. If not, a default SnapshotClass will be used.
+
+* For statically binding snapshot, user/admin must specify the pointers in both directions (VolumeSnapshot → VolumeSnapshotData, and VolumeSnapshotData → VolumeSnapshot) so that controller will bind them.
```

Comment:
"so that controller will bind them" - if both pointers are set, what exactly will the controller do to "bind them"? In PV-PVC case, these pointers *are* the binding and PV controller won't do anything when both pointers are set (except updating PV/PVC status to "Bound").

Reply:
Ok.  This will be re-worded.